rep-2011.rst

```diff
@@ -0,0 +1,1421 @@
+
+Another use case, which is not covered in this REP, is using the type description in tooling to either display the type description to the user or to include it in recordings.
+
+In either case, where the type description comes from doesn't really matter, and so, for example, it could be looked up on the local filesystem or read from a rosbag file.
+However, in practice, the correct type description may not be found locally, especially in cases where you have different versions of messages in the same system, e.g.:
+
+- because it's on another computer, or
+- because it is from a different distribution of ROS, or
+- because it was built in a different workspace, or
+- because the application has not been restarted since recompiling a change to the type being used
+
+In any case, it is useful to have a mechanism to convey the type descriptions from the source of the data to other nodes, which we describe here as "type description distribution".
+
+Furthermore, this feature should be agnostic to the underlying middleware and serialization library, as two endpoints may not have the same rmw implementation, or the data may have been serialized to a different format in the case of playback of a recording.
+
+Sending the Type Description
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Type descriptions will be provided by a ROS Service called ``~/_get_type_description``, which will be offered by each node.
+There will be a single ROS Service per node, regardless of the number of publishers or subscriptions on that node.
+
+This service may be optional, for example being enabled or disabled when creating the ROS Node.
+
+.. TODO::
+
+    How can we detect when a remote node is not offering this service?
+    It's difficult to differentiate between "the Service has not been created yet, but will be" and "the Service will never be created".
+    Should we use a ROS Parameter to indicate this?
+    But then what if remote access to Parameters (another optional Service) is disabled?
+    Perhaps we need a "services offered" list which is part of the Node metadata, which is sent for each node in the rmw implementation, but that's out of scope for this REP.
+
+A service request to this ROS Service will comprise of the type name and the type version hash, which is distributed during discovery of endpoints and will be accessible through the ROS Graph API, as described in previous sections.
+The ROS Service server will respond with the type description and any necessary metadata needed to do Run-Time Interface Reflection.
+This service is not expected to be called frequently, and is likely to only occur when new topic or service endpoints are created, and even then, only if the endpoint type hashes do not match.
+
+.. TODO::
+
+    Should each endpoint be asked about their type/version pair or should we assume that the type/version pair guarantees a unique type description and therefore reuse past queries?
```

Comment:
It seems to me best to make a strong statement that type/version pair is a unique type description so a user can expect that anybody with that name and hash has the same type. From bag recording perspective, with potentially a lot of topics, it'd be nice to only call out once per unique type, and I'd probably just base it on hash lookup in internal definition cache, whether to call or not